File: scripts/ncbiTaxonomyParser.pl
#!/usr/bin/env perl
use strict;
use warnings;
use Cwd;
use Getopt::Std;
use IO::Handle;
use File::Path;

=desc
parse info from taxonomy NCBI dmp files (names.dmp and nodes.dmp)
to create a mapping file taxonNamesFull.txt that contains
ncbiID	fullName	rank	parentID
23.01.2017
=cut

sub usage {
    my $msg = shift;
    print "example: perl ncbiTaxonomyParser.pl -i dmp_folder\n";
    print "-i\tFolder contains DMP files (names.dmp and nodes.dmp)\n";
    die $msg."\n";
}

# global variables
our($opt_i);
getopts('i:');

# sanity checks;
my $input = ($opt_i) ? $opt_i : usage("ERROR: No input folder given\n");

### parse names and IDs from names.dmp
print "Parsing names.dmp...";

open(NAME,"$input/names.dmp") || die "Cannot open $input/names.dmp!\n";
my @nameIN = <NAME>;
close (NAME);

my %name;

foreach my $line(@nameIN){
	chomp($line);
	if($line =~ /scientific name/){
		my @tmp = split(/\t\|\t/,$line);	# taxID - name
		my $name = $tmp[1];
		$name =~ s/\'//g;
		$name{$tmp[0]} = $name;
	}
}
print " Finished!\n";

### parse IDs and parents IDs
print "Parsing nodes.dmp...";

open(NODE,"$input/nodes.dmp") || die "Cannot open $input/nodes.dmp!\n";
my @nodeIN = <NODE>;
close (NODE);

open(OUT,">$input/taxonNamesFull.txt");
print OUT "ncbiID	fullName	rank	parentID\n";

my $c = 0;
foreach my $line(@nodeIN){
	chomp($line);
	my @tmp = split(/\t\|\t/,$line);	# taxID - parentID - rank

	## output: ncbiID	fullName	rank	parentID
	my $rank = $tmp[2];
	$rank =~ s/\s//g; $rank =~ s/\'//g;
	print OUT "$tmp[0]\t$name{$tmp[0]}\t$rank\t$tmp[1]\n";

	$c ++;
	print $c,"/",scalar(@nodeIN),"\n";
}
print " Finished!\n Check your output at\n$input/taxonNamesFull.txt\n";
close (OUT);

exit;

File: inst/PhyloProfile/scripts/ncbi_taxonomy_parser.py
# -*- coding: utf-8 -*-
#!/usr/bin/env python

# Parse info from taxonomy NCBI dmp files (names.dmp and nodes.dmp)
# to create a mapping file taxonNamesFull.txt that contains
# ncbiID	fullName	rank	parentID
# Author: Carla Mölbert {carla.moelbert@gmx.de}
# Date: 10.07.2018

import argparse
import subprocess

# Get the arguments for the script
parser = argparse.ArgumentParser()
parser.add_argument('-i', '--input',
					help = 'folder contain the names.dmp and nodes.dmp files(required)',
					required = True)
args = parser.parse_args()

# Generate dictionary with key = ncbiID and value = fullName
def get_names_dict(names):
	names_dict = dict()
	for name in names:
		if ("scientific name" in name):
			tmp = name.split("\t|\t")
			names_dict[tmp[0]] = tmp[1]
	return(names_dict)

# Get a new line for the output file
def get_new_line (node, names_dict):
	tmp = node.split("\t|\t")
	ncbiID = tmp[0]
	parentID = tmp[1]
	rank = tmp[2]
	rank = rank.replace("'", "")
	rank = rank.replace(" ", "")
	fullName = names_dict[ncbiID]
	fullName = fullName.replace("'", "")
	new_line = ncbiID + "\t" + fullName + "\t" + rank + "\t" + parentID + "\n"
	return(new_line)

########################################
output = args.input + "taxonNamesFull.txt"
file = open(output,"w")
file.write("ncbiID\tfullName\trank\tparentID\n") # header

# names to open the files
filename_names = args.input + "/names.dmp"
filenames_nodes = args.input + "/nodes.dmp"

# Get the needed information from names.dmp
with open(filename_names) as names_file:
	print("Read names.dmp \n")
	names = names_file.readlines()
names_dict = get_names_dict(names)

# Combine the information with the information from nodes.dmp
with open (filenames_nodes) as nodes_file:
	print("Read nodes.dmp \n")
	nodes = nodes_file.read().splitlines()

print("Write taxonNamesFull.txt")
for node in nodes:
	new_line = get_new_line(node, names_dict)
	file.write(new_line)

msg = "Finished! Please replace phyloprofile/data/taxonNamesFull.txt by \n" + output + "!"
print(msg)
